src/jit/gentree.cpp

```diff
@@ -6586,115 +6799,165 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
+        }
+    }
+    GenTreeBlk* newBlkOrObjNode = new (this, GT_OBJ) GenTreeObj(nodeType, addr, structHnd, size);
+
     // An Obj is not a global reference, if it is known to be a local struct.
-    GenTreeLclVarCommon* lclNode = addr->IsLocalAddrExpr();
-    if ((lclNode != nullptr) && !lvaIsImplicitByRefLocal(lclNode->gtLclNum))
+    if ((addr->gtFlags & GTF_GLOB_REF) == 0)
     {
-        objNode->gtFlags &= ~GTF_GLOB_REF;
+        GenTreeLclVarCommon* lclNode = addr->IsLocalAddrExpr();
+        if ((lclNode != nullptr) && !lvaIsImplicitByRefLocal(lclNode->gtLclNum))
+        {
+            newBlkOrObjNode->gtFlags &= ~GTF_GLOB_REF;
+        }
     }
-    return objNode;
+    return newBlkOrObjNode;
 }
 
-// Creates a new CpObj node.
-// Parameters (exactly the same as MSIL CpObj):
+//------------------------------------------------------------------------
+// gtSetObjGcInfo: Set the GC info on an object node
 //
-//  dst        - The target to copy the struct to
-//  src        - The source to copy the struct from
-//  structHnd  - A class token that represents the type of object being copied. May be null
-//               if FEATURE_SIMD is enabled and the source has a SIMD type.
-//  isVolatile - Is this marked as volatile memory?
-GenTreeBlkOp* Compiler::gtNewCpObjNode(GenTreePtr dst, GenTreePtr src, CORINFO_CLASS_HANDLE structHnd, bool isVolatile)
-{
-    size_t    size       = 0;
-    unsigned  slots      = 0;
-    unsigned  gcPtrCount = 0;
-    BYTE*     gcPtrs     = nullptr;
-    var_types type       = TYP_STRUCT;
-
-    GenTreePtr hndOrSize = nullptr;
+// Arguments:
+//    objNode - The object node of interest
 
-    GenTreeBlkOp* result = nullptr;
+void Compiler::gtSetObjGcInfo(GenTreeObj* objNode)
+{
+    CORINFO_CLASS_HANDLE structHnd  = objNode->gtClass;
+    var_types            nodeType   = impNormStructType(structHnd);
+    unsigned             size       = objNode->gtBlkSize;
+    unsigned             slots      = 0;
+    unsigned             gcPtrCount = 0;
+    BYTE*                gcPtrs     = nullptr;
 
-    bool useCopyObj = false;
+    assert(varTypeIsStruct(nodeType));
+    assert(size == info.compCompHnd->getClassSize(structHnd));
 
-    // Intermediate SIMD operations may use SIMD types that are not used by the input IL.
-    // In this case, the provided type handle will be null and the size of the copy will
-    // be derived from the node's varType.
-    if (structHnd == nullptr)
+    if (nodeType == TYP_STRUCT)
     {
-#if FEATURE_SIMD
-        assert(src->OperGet() == GT_ADDR);
-
-        GenTree* srcValue = src->gtGetOp1();
+        if (size >= TARGET_POINTER_SIZE)
+        {
+            // Get the GC fields info
+            var_types simdBaseType; // Dummy argument
+            slots    = (unsigned)(roundUp(size, TARGET_POINTER_SIZE) / TARGET_POINTER_SIZE);
+            gcPtrs   = new (this, CMK_ASTNode) BYTE[slots];
+            nodeType = impNormStructType(structHnd, gcPtrs, &gcPtrCount, &simdBaseType);
```

Comment:
This is the first/only place where we get the actual GC layout for object nodes (the other place we do this for struct types is when we set up the lclVars). Prior to this, it is necessary that we know the type and size, so we have to call `impNormStructType()` without getting the GC layout. We may determine that it is better to pay the full price up-front, but that leads to a different set of JIT-EE calls than we were previously doing.

---
In reply to: [77385804](https://github.com/dotnet/coreclr/pull/7027#discussion_r77385804) [](ancestors = 77385804)